proof/courier.go

```diff
@@ -90,20 +55,123 @@ type CourierHarness interface {
 	// information in the Addr type.
-	DeliverProof(context.Context, Addr, *AnnotatedProof) error
+	DeliverProof(context.Context, Recipient, *AnnotatedProof) error
 
 	// ReceiveProof attempts to obtain a proof as identified by the passed
 	// locator from the source encapsulated within the specified address.
-	ReceiveProof(context.Context, Addr, Locator) (*AnnotatedProof, error)
+	ReceiveProof(context.Context, Recipient, Locator) (*AnnotatedProof,
+		error)
 
 	// SetSubscribers sets the set of subscribers that will be notified
 	// of proof courier related events.
 	SetSubscribers(map[uint64]*fn.EventReceiver[fn.Event])
 }
 
+// CourierAddr is a fully validated courier address (including protocol specific
+// validation).
+type CourierAddr interface {
+	// Url returns the url.URL representation of the courier address.
+	Url() *url.URL
+
+	// NewCourier generates a new courier service handle.
+	NewCourier(ctx context.Context, cfg *CourierCfg) (Courier, error)
+}
+
+// ParseCourierAddrString parses a proof courier address string and returns a
+// protocol specific courier address instance.
+func ParseCourierAddrString(addr string) (CourierAddr, error) {
+	// Parse URI.
+	urlAddr, err := url.ParseRequestURI(addr)
+	if err != nil {
+		return nil, fmt.Errorf("invalid proof courier URI address: %w",
+			err)
+	}
+
+	return ParseCourierAddrUrl(*urlAddr)
+}
+
+// ParseCourierAddrUrl parses a proof courier address url.URL and returns a
+// protocol specific courier address instance.
+func ParseCourierAddrUrl(addr url.URL) (CourierAddr, error) {
+	// Create new courier addr based on URL scheme.
+	switch addr.Scheme {
+	case ApertureCourier:
+		return NewHashMailCourierAddr(addr)
+	}
+
+	return nil, fmt.Errorf("unknown courier address protocol: %v",
+		addr.Scheme)
+}
+
+// HashMailCourierAddr is a hashmail protocol specific implementation of the
+// CourierAddr interface.
+type HashMailCourierAddr struct {
+	addr url.URL
+}
+
+// Url returns the url.URL representation of the hashmail courier address.
+func (h *HashMailCourierAddr) Url() *url.URL {
+	return &h.addr
+}
+
+// NewCourier generates a new courier service handle.
+func (h *HashMailCourierAddr) NewCourier(_ context.Context,
+	cfg *CourierCfg) (Courier, error) {
+
+	hashMailBox, err := NewHashMailBox(&h.addr, cfg.HashMailCfg.TlsCertPath)
+	if err != nil {
+		return nil, fmt.Errorf("unable to make mailbox: %v",
+			err)
+	}
+
+	subscribers := make(
+		map[uint64]*fn.EventReceiver[fn.Event],
+	)
+	return &HashMailCourier{
+		cfg:         cfg.HashMailCfg,
+		mailbox:     hashMailBox,
+		deliveryLog: cfg.DeliveryLog,
+		subscribers: subscribers,
+	}, nil
+}
+
+// NewHashMailCourierAddr generates a new hashmail courier address from a given
+// URL. This function also performs hashmail protocol specific address
+// validation.
+func NewHashMailCourierAddr(addr url.URL) (*HashMailCourierAddr, error) {
+	// We expect the port number to be specified for a hashmail service.
+	if addr.Port() == "" {
```

Comment:
I think it should also verify aspects such as the `Scheme` (should be hashmail in this case), etc.

Reply:
Ok, no problem. I can move verification of all properties into protocol specific new address functions.